### ISSUE TITLE
Rewrite e2e tests for `overfunded_htlc` and `underfunded_htlc`

### ISSUE THREAD

**Issue by bonomat** [comit-network/comit-rs#1867]
## Problem

Some e2e have been rewritten already #1584/#1771 following the new design outlined in [spike 11](https://github.com/comit-network/spikes/blob/master/0011-revise-e2e-test-framework.adoc#the-ideal-test).

## Goal

Use the new design for `overfunded_htlc` and `underfunded_htlc` as well.

## Recommendation

Note: To do this, we will have to do something about the problem identified by @bonomat here: https://github.com/comit-network/comit-rs/issues/1594#issuecomment-559624013

To move forward, the best way seems to be to just make those methods public as stated by @bonomat.

**Comment by thomaseizinger** [comment]
To do this, we will have to do something about the problem identified by @bonomat here: https://github.com/comit-network/comit-rs/issues/1594#issuecomment-559624013

To move forward, the best way seems to be to just make those methods public as stated by @bonomat.

**Comment by da-kami** [comment]
I just re-discussed this with @D4nte and we came to the conclusion that the proposed solution of exposing the functions in the SDK is not a good way. We would expose it solely to test un-wanted behaviour in the e2e tests. 

This solution was also not the preferred solution of the discussion in the original ticket https://github.com/comit-network/comit-rs/issues/1594

I see 2 different concerns here:
1. SDK for Applications Devs.
2. SDK for our e2e tests.

Where should we draw the line for changes that concern the e2e tests?

My recommendation: 

1. Do not expose SDK functions
2. Re-write the e2e test...
    1. extend the actors' `sendRequest()` with optional quantity (so I can control through test how much quantity is in the request, if parameter not given fallback to defaults like now)
    2. extend `lib_sdk/wallets/bitcoin.ts` and `lib_sdk/wallets/ethereum.ts` to allow sending transactions to the wallet.
    3. add function `fund_with_quantity(quantity: string)` to the `lib_sdk` in e2e tests that:
        1. keeps calling `fetchDetails()` of the swap till `fund` action is available
        2. funds with the given quantity directly against the wallet instead of using the fund action of the SDK.

This, of course means that we duplicate some code (basically re-implement `Swap.executeAction()` and parts of `doLedgerAction()` in the `lib_sdk` of the e2e tests. But I feel this is cleaner.

**Comment by bonomat** [comment]
I think the functions we want to test here (over-/underfudned) are edge cases and we should try to not change the default behavior too much, i.e. not adding the proposed extensions to `lib_sdk/*`. 

I have a counter proposal based on the idea I followed in #1978 which might even be simpler:

The idea is based on having standalone tests which doe everything themselves, e.g.:

* implement a custom `BitcoinWallet` only for this test which simply sends 5x or 0.5x of the money it's supposed to send (c&p) the `InMemoryWallet` from the sdk and change `sendToAddress()`
* create two `ComitClient`s for Alice and Bob (one with the OverfundingBitcoinWallet)
* send a swap request to bob via `aliceComitClient.sendSwap()` 
* call default funding function which should then invoke the overfunding `sendToAddress` function on the wallet (i.e. `swap.fund()`)
* wait and assert the wrong state on both cnd nodes


You can do the same thing for overfunding ethereum/erc20 etc.


I feel like this also reflects the real world behaviour, i.e. the user does a mistake while sending the funds.

**Comment by thomaseizinger** [comment]
So @bonomat and I just discussed this.

1. https://github.com/comit-network/comit-js-sdk/issues/94 is likely going to remove `fund` etc from the `Swap` interface. Given that, those functions will need to be public anyway.
2. Even without https://github.com/comit-network/comit-js-sdk/issues/94, I don't consider our SDK to be in a state where we need to use velvet gloves (as we like to say in German :P). Just change it to whatever we need :D 
3. IMO, from the test's PoV, the difference between a regular test and one that overfunds should be at the `fund` step. @bonomat's proposal doesn't cover that as the difference is only in the setup.

I would proceed as follows:

1. Make the functions in the SDK public.
2. Make `Actor.fund` have an optional parameter for overriding the amount.
3. Don't use `Swap.fund` inside `Actor.fund` but do whatever these functions do and manipulate the response before sending it to the wallet.

Retrospectively, I would consider it a mistake to add the `fund` etc functions to the SDK. We optimized too early for what should go in the SDK :)
These functions should only exist in the e2e tests because those rightfully must have an opinion on what the next action should be in order to verify cnd's behaviour.
